Validate bare values

RobinMalfait opened the pull request:
This PR validates incoming bare values to ensure we generate CSS that makes sense.

Before this PR, we would generate values such as:

```css
.z-index {
  z-index: index;
}

.flex-foo {
  flex: foo;
}
```

Which is invalid, after this PR, these vaues would not be generated at all.

<!--

👋 Hey, thanks for your interest in contributing to Tailwind!

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a lot of time and effort into a new feature. To avoid this from happening, we request that contributors create an issue to first discuss any significant new features. This includes things like adding new utilities, creating new at-rules, or adding new component examples to the documentation.

https://github.com/tailwindcss/tailwindcss/blob/master/.github/CONTRIBUTING.md

-->

RobinMalfait committed:
add validation for bare values

RobinMalfait committed:
add bare value validation to `row-span`, `row-start`, `row-end`, `row-span`, `row-start`, and `row-end`

RobinMalfait committed:
update changelog

RobinMalfait commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -628,7 +634,10 @@ export function createUtilities(theme: Theme) {
   functionalUtility('col-span', {
     themeKeys: [],
-    handleBareValue: ({ value }) => value,
+    handleBareValue: ({ value }) => {
```
I put the validation for `col-span`, `col-start`, `col-end`, `row-span`, `row-start`,  and `row-end` in a separate commit because they can technically almost contain any value. If a "name" is used, then they would refer to named grid areas: https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
```css
/* <custom-ident> value */
grid-column-start: somegridarea;

/* <integer> + <custom-ident> values */
grid-column-start: 2;
grid-column-start: somegridarea 4;
```

So the question is, do we want to allow anything here, or is it fine to only accept integer values for bare values, e.g.: `col-span-123` and if you want to generate values using names that you should use arbitrary values instead?

adamwathan replied:
I kinda think just do numbers and use arbitrary values for other stuff 👍🏻

thecrypticace replied:
Agree, I think bare values should just be numbers. Feels much simpler to explain that too.

RobinMalfait replied:
Alright perfect, than I can keep the commit in there, nice 👍

thecrypticace commented:
This looks good. I'm just thinking out loud but I wonder if we can lift this check up a level to before we call `handleBareValue`? The only oddball out then is `aspect-*` in which the bare value is `{number}/{number}` but I'm not sure how we'd want to handle that. Any ideas?

adamwathan reviewed (commented):
Looks good so far, I think missing handling the `flex` root though, because current alphas generate this for example:

```css
.flex-direction {
  flex: direction;
}
```

RobinMalfait committed:
validate bare values in `utilities.functional` utilities

RobinMalfait committed:
mark candidates as invalid if the leftover value is an empty string

thecrypticace commented on packages/tailwindcss/src/candidate.ts:
```diff
@@ -426,6 +426,13 @@ export function parseCandidate(input: string, designSystem: DesignSystem): Candi
 
+      // If the leftover value is an empty string, it means that the value is an
+      // invalid named value. This makes the candidate invalid and we can
+      // skip continue parsing it.
```
```suggestion
      // skip any further parsing
```

thecrypticace reviewed (changes requested):
one small nit about a comment but otherwise looks good!

RobinMalfait committed:
Update packages/tailwindcss/src/candidate.ts

Co-authored-by: Jordan Pittman <jordan@cryptica.me>

RobinMalfait committed:
validate `from`, `via` and `to`

This way we make sure that the bare value ends with `%`, and the value
before it is a number.
